lib/utils/classes.py

```diff
@@ -21,3 +26,157 @@ class Stock:
+        self.__session.get(self.__login_url)
+        csrftoken = self.__session.cookies['csrftoken']
+
+        #Appending the csrftoken to the data payload
+        login_payload = {'username':self.__username, 'password':self.__password, 'csrfmiddlewaretoken':csrftoken}
+        resp = self.__session.post(self.__login_url, data = login_payload, headers=dict(Referer=self.__login_url))
+        
+    def mySession(self) -> [requests.sessions.Session]:
+        return self.__session
+
+    def closeSession(self):
+        self.__session.close()
+
+
+
+class Companies:
+    """
+    Usage:
+    Companies = Companies(session, 101)
+    myCompanies = Companies.getCompanyTickers()
+    """
+    __screen_url: str = 'https://www.screener.in/screen/raw/'
+    __screen_referer_url: str = 'https://www.screener.in/screen/new/'
+    __company_search_url = 'https://www.screener.in/api/company/search/'
+
+    __limit: int
+
+    __session: [requests.sessions.Session]
+
+    def __init__(self,session,limit=100000):
+        self.__limit = limit
+        self.__session = session
+        #self.__companies = pd.DataFrame() #empty DF to append to
+        
+    def setLimit(self,limit=100000, refetch = False):
+        self.__limit = limit
+        if(refetch == True):
+            self.__fetchCompanies()
+
+    def __fetchCompanies(self):
+        screen_referer_url = self.__screen_referer_url
+        screen_payload = {'sort':'','source':'','query': 'Current price '}
+
+        companies_df = pd.DataFrame() #empty DF to append to
+
+        page_iter = 1
+        row_count = -1
+        while(True):
+            # Get the webpage after submitting the query to screen page
+            screen_page = self.__session.get(self.__screen_url, params = screen_payload, headers=dict(Referer=screen_referer_url))
+            #print(screen_page.url) #Prints the URL of the new web page
+
+            # Extract the table into dataframe. skiprows is used to remove the un necessary rows
+            table = pd.read_html(screen_page.text, index_col = 0, header = 0, skiprows = [16, 32, 48])[0] #table is a list
+            no_of_rows = table.shape[0]
+
+            # Updating the row_count using the number of rows from 1st page
+            if(row_count == -1):
+                row_count = no_of_rows
+                if(row_count == 0):
+                    return #TODO: Need to add a error here
+            
+            companies_df = pd.concat([companies_df, table], ignore_index = False)
+            page_iter = page_iter + 1
+
+            screen_referer_url = screen_page.url
+            if(no_of_rows != row_count):
+                break
+            elif(companies_df.shape[0] >= self.__limit):
+                companies_df = companies_df.loc[companies_df.index <= (self.__limit)]
+                break
+            else:
+                screen_payload['page'] = str(page_iter)
+            sleep(1)
+        companies_df.to_csv('../../Companies.csv')
```

Comment:
Why do we need to save this as a CSV? The file's interface was to merely return an array of ticker symbols

If you just want a way to access it later, store it in memory - it's not a huge amount of data 😄

Reply:
It serves 2 purposes. 
1. We don't need to refetch it again
2. Modify this CSV to include additional columns like "Analysed", "Study further", etc

Reply:
Well the problem with that approach is that your information is now stateful - which means it depends on this CSV database

this also makes the code less maintainable in the long run because instead of modularising, everything will end up being in one large file. Having different files with individual small functions helps prevent bugs and makes everything expandable which is why I created the Microstructure 😃 

not having a database also enables us to run this wherever we want 


we also probably won’t fetch all information together and do it in batches because screener will block us if we send too many requests too quickly - so in this case it’s probably better to fetch things separately

Reply:
How do you plan to track which tickers have already been fetched and analyzed?

Reply:
We don't have to - at least not at this stage
If we try and solve the entire problem at once, we'll make compromises and spaghetti code :( 
We need to break everything down to simple functions and implement those functions independently

Which is why I made this structure:

```
└── stock-analyser
    ├── README.md 
    ├── cli => Runnable file
    └── lib => Source code directory
        ├── analyser.py => Given a Stock object, analyse it
        ├── fetcher.py => Fetches a list of stocks, somehow, and returns just the list
        ├── scraper.py => Given a list of stock tickers, scrape related data and store in a Stock object
        └── utils => Utilities ==> *MUST* not contain main source code - these are just helpers
        │   ├── args.py => Parse arguments from command line
        │   ├── classes.py => Contains the definition of Stock, Document => Doesn't do anything itself
        │   └── version_check.py => Makes sure we're running the right python version
        └── zipper.py => Convert files to .ZIP to export for manual review
```

Your implementation is great - it works - but it doesn't conform to this structure because you have all the main code in the utils/classes.py file, and the fetcher.py file does absolutely nothing :) Running the CLI also does nothing

If you don't like this structure - we can change it - but let's not have a structure that we're not using.